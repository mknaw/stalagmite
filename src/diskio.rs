use std::path::Path;
use std::sync::Arc;

use camino::{Utf8Path, Utf8PathBuf};
use ignore::Walk;
use memmap2::Mmap;

use crate::common::{RenderRules, SiteEntry, SiteNode, DEFAULT_RENDER_RULE_SET};
use crate::Config;

/// Read a `path` to an `Mmap`.
pub fn read_file_contents<P: AsRef<Path>>(path: P) -> Mmap {
    let file = std::fs::File::open(path).unwrap();
    unsafe { Mmap::map(&file).unwrap() }
}

/// Recursively iterate over all files in the given directory with the given extension.
pub fn walk<'a, P: AsRef<Path>>(
    dir: P,
    ext: &'a Option<&'a str>,
) -> Box<dyn Iterator<Item = Utf8PathBuf> + 'a> {
    let walk = Walk::new(dir).flatten().filter_map(|entry| {
        let path = entry.path();
        if path.is_file() {
            match ext.as_ref() {
                Some(ext) => {
                    if path.extension() == Some(ext.as_ref()) {
                        Some(Utf8PathBuf::from_path_buf(path.to_owned()).unwrap())
                    } else {
                        None
                    }
                }
                None => Some(Utf8PathBuf::from_path_buf(path.to_owned()).unwrap()),
            }
        } else {
            None
        }
    });
    Box::new(walk)
}

/// Recursively iterate over pages directory, sending data pertaining to each directory to a `sink`.
/// Technically the site topology is a tree, but currently have no need to represent it as such.
pub fn collect_site_nodes(config: Arc<Config>) -> Vec<SiteNode> {
    // TODO probably could do less hand rolling if we just walkdir, get directories, and then
    // output a `SiteNode` for each of those.
    fn recurse(
        site_nodes: &mut Vec<SiteNode>,
        pages_dir: &Utf8Path,
        current_path: &Utf8Path,
        rules_stack: &mut Vec<Arc<RenderRules>>,
    ) {
        let rules_path = current_path.join("rules.yaml");
        if rules_path.exists() && rules_path.is_file() {
            let raw = std::fs::read_to_string(&rules_path).unwrap();
            // TODO wanted to override the previous rules on the stack...
            // TODO might not need Arc anymore, now that SiteNodes own the RuleSets?
            let rule_set: Arc<RenderRules> = Arc::new(serde_yaml::from_str(&raw).unwrap());
            rules_stack.push(rule_set);
        }

        if let Ok(dir_entries) = std::fs::read_dir(current_path) {
            let paths: Vec<Utf8PathBuf> = dir_entries
                .filter_map(|entry| {
                    entry
                        .ok()
                        .map(|e| Utf8PathBuf::from_path_buf(e.path()).unwrap())
                })
                .collect();
            let site_entries: Vec<SiteEntry> = paths
                .iter()
                .filter_map(|path| SiteEntry::try_new(pages_dir, path.clone()).ok())
                .collect();

            if !site_entries.is_empty() {
                site_nodes.push(SiteNode {
                    dir: current_path.strip_prefix(pages_dir).unwrap().to_path_buf(),
                    render_rules: rules_stack.last().expect("rules_stack is empty").clone(),
                    site_entries,
                });
            }

            paths.iter().filter(|path| path.is_dir()).for_each(|path| {
                recurse(site_nodes, pages_dir, path, rules_stack);
            });
        }

        // Pop off the stack when backtracking
        if rules_path.exists() && rules_path.is_file() {
            rules_stack.pop();
        }
    }

    let mut site_nodes: Vec<SiteNode> = Vec::new();
    let mut rules_stack = vec![DEFAULT_RENDER_RULE_SET.clone()];
    let pages_dir = config.pages_dir();
    recurse(&mut site_nodes, &pages_dir, &pages_dir, &mut rules_stack);
    site_nodes
}

// TODO this P, P2 thing is unseemly.
// TODO should callers just take care of joining the path?
// TODO probably also want to minify + compress
pub async fn write_html<P: AsRef<Path>>(out_path: P, html: &str) -> anyhow::Result<()> {
    tracing::debug!("Writing HTML to {}", out_path.as_ref().display());
    tokio::fs::create_dir_all(out_path.as_ref().parent().unwrap()).await?;
    tokio::fs::write(&out_path, html).await?;
    Ok(())
}

// TODO what color is my function?
pub fn write_html_sync<P: AsRef<Path>>(out_path: P, html: &str) -> anyhow::Result<()> {
    tracing::debug!("Writing HTML to {}", out_path.as_ref().display());
    std::fs::create_dir_all(out_path.as_ref().parent().unwrap())?;
    std::fs::write(&out_path, html)?;
    Ok(())
}
