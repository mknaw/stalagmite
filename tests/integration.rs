use std::sync::Arc;

use camino::Utf8PathBuf;
use stalagmite::diskio::walk;
use stalagmite::{generate, Config};

#[tokio::test]
async fn generate_example_site() {
    let example_project_dir =
        Utf8PathBuf::from_path_buf(std::env::current_dir().unwrap().join("example")).unwrap();

    std::fs::remove_dir_all(example_project_dir.join("public")).ok();
    // TODO this is not the one it ends up using...
    std::fs::remove_file(example_project_dir.join("db.sqlite")).ok();

    let config = Arc::new(Config::init(Some(example_project_dir.clone())).unwrap());

    // TODO should have an in-memory sqlite for testing.
    generate(config).await.unwrap();

    let out_dir = example_project_dir.join("public");
    let mut files: Vec<Utf8PathBuf> = walk(&out_dir, &Some("html"))
        .map(|p| p.strip_prefix(&out_dir).unwrap().to_owned())
        .collect();
    files.sort();
    assert_eq!(
        files,
        vec![
            "blog/0/index.html",
            "blog/welcome-to-my-blog/index.html",
            "index.html",
        ]
    );

    for file in walk(example_project_dir.join("public"), &Some("html")) {
        let contents = std::fs::read_to_string(&file).unwrap();
        insta::assert_yaml_snapshot!(contents.split('\n').collect::<Vec<&str>>());
    }
}
