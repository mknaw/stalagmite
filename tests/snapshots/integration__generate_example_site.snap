---
source: tests/integration.rs
expression: "contents.split('\\n').collect::<Vec<&str>>()"
---
- "<!DOCTYPE html>"
- "<html lang=\"en\">"
- "    <head>"
- "        <meta charset=\"UTF-8\">"
- "        <meta name=\"viewport\" content=\"width=device-width, initial-scale=1.0\">"
- "        <title>Example Stalagmite Site</title>"
- "        <link rel=\"stylesheet\" href=\"/static/tw.css\">"
- "        <script src=\"assets/dev-reload.26ccffdaf8b45deb.js\"></script>"
- "    </head>"
- "    <body>"
- "        <main>"
- "            <nav class=\"bg-gray-800\">"
- "                <div class=\"container mx-auto px-6 py-3\">"
- "                    <div class=\"flex justify-between items-center\">"
- "                        <div class=\"flex\">"
- "                            <div class=\"text-white text-lg font-semibold\">Stalagmite</div>"
- "                            <div class=\"ml-10 space-x-4\">"
- "                                <a href=\"#home\" class=\"text-gray-300 hover:text-white\">Home</a>"
- "                                <a href=\"#blog\" class=\"text-gray-300 hover:text-white\">Blog</a>"
- "                                <a href=\"#projects\" class=\"text-gray-300 hover:text-white\">Projects</a>"
- "                            </div>"
- "                        </div>"
- "                    </div>"
- "                </div>"
- "            </nav>"
- ""
- "            <div class=\"bg-blue-500\">Home page is here!</div>"
- ""
- "        </main>"
- "    </body>"
- "</html>"

